### PR TITLE
test(cli): Add cross env tests for `gen,sync` commands

### PR DESCRIPTION
**Pull request by erezrokah**
#### Summary

Adds basic integration tests. The `sync` test is failing due to a bug in our release pipeline (not embedding versions) which I'll fix in a separate PR

---

Use the following steps to ensure your PR is ready to be reviewed

- [ ] Read the [contribution guidelines](../blob/main/CONTRIBUTING.md) 🧑‍🎓
- [ ] Test locally on your own infrastructure
- [ ] Run `go fmt` to format your code 🖊
- [ ] Lint your changes via `golangci-lint run` 🚨 (install golangci-lint [here](https://golangci-lint.run/usage/install/#local-installation)) 
- [ ] Update or add tests 🧪
- [ ] Ensure the status checks below are successful ✅

### CODE REVIEW

File: .github/workflows/test_build.yml [hunk 1]
```diff
@@ -19,7 +18,7 @@ jobs:
         uses: actions/checkout@v3
       - id: set-matrix
         run: ./scripts/resolve-modules.sh
-  test_build:
+  test-build:
```
Comment (erezrokah) [inline .github/workflows/test_build.yml:21]: Keep casing consistent

File: .github/workflows/test_build.yml [hunk 2]
```diff
@@ -6,7 +6,6 @@ on:
       - main
 env: 
   CGO_ENABLED: 0
-  CQ_NO_TELEMETRY: 1
```
Comment (erezrokah) [inline .github/workflows/test_build.yml:9]: I don't think we need this anymore

File: .github/workflows/test_integration.yml [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+on:
+  pull_request:
+    branches:
+      - main
+env: 
+  CGO_ENABLED: 0
+
+jobs:
+  check-cli-changes:
```
Comment (erezrokah) [inline .github/workflows/test_integration.yml:11]: This should save spinning up the `windows,linux,macOS` matrix if the CLI hadn't changed

File: cli/integration_tests/gen_test.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+func TestGen(t *testing.T) {
+	for _, plugin := range pluginsToTest {
+		t.Run(fmt.Sprintf("gen %s %s", plugin.kind, plugin.name), func(t *testing.T) {
+			i := newIntegrationTest(t)
+			output, err := i.runCommand(t, "gen", plugin.kind, plugin.name)
+			fmt.Println(output)
+			require.NoError(t, err)
+			require.Contains(t, output, "plugin config successfully written to "+plugin.name+".yml")
+			config := validateYaml(t, path.Join(i.dir, plugin.name+".yml"))
+			require.Equal(t, plugin.kind, config.Kind)
```
Comment (erezrokah) [inline cli/integration_tests/gen_test.go:45]: This is a sanity check. We shouldn't be testing the whole parsing here

File: cli/integration_tests/sync_test.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"path"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func addDestination(t *testing.T, configFile string) {
```
Comment (erezrokah) [inline cli/integration_tests/sync_test.go:15]: We should be able to remove this once https://github.com/cloudquery/cloudquery/pull/1847 is merged

**Comment by erezrokah** [comment]
OK so I found https://github.com/cloudquery/cloudquery/blob/9e3ace775da2d600968ef4275e9e0013d4dfd825/cli/cmd/generate_test.go#L70 and https://github.com/cloudquery/cloudquery/blob/9e3ace775da2d600968ef4275e9e0013d4dfd825/cli/cmd/sync_test.go#L9 so I'll merge the too.
The separation to integration tests is so we can run them on Windows/MacOS too

**Comment by erezrokah** [comment]
OK so this PR improves on the current test by:
1. Runs each test in a separate isolated directory
2. Runs the CLI via an external process closer to how the user runs int
3. Separate the integration tests from the unit tests and runs them on a multiple OS matrix

**Comment by erezrokah** [comment]
Looks like the CLI v2 doesn't work on Windows https://github.com/cloudquery/cloudquery/actions/runs/3077812862/jobs/4972939978#step:7:42

**Comment by yevgenypats** [comment]
> OK so this PR improves on the current tests by:
> 
> 1. Runs each test in a separate isolated directory so easy to add more scenarios
> 2. Runs the CLI via an external process closer to how the user runs it
> 3. Separate the integration tests from the unit tests and runs them on a multiple OS matrix

Im not a big fan of running this test as an external process as it makes testing good and bad test cases much hard as it will involve stdout output. I specifically built the cobra in such way that it should be possible to do so via `NewRootCmd` which is the main entry point so there shouldn't be any difference between this and an external process.

**Comment by erezrokah** [comment]
> I specifically built the cobra in such way that it should be possible to do so via `NewRootCmd` which is the main entry point so there shouldn't be any difference between this and an external process.

Is there a way to run isolated tests in a temp directory and not in the CWD? That should allow us to reduce dependencies between tests and run tests in parallel.

Also I think testing stdout is not necessarily bad in the case of CLIs as that's what the user experiences. We can leverage snapshot testing on the CLI output to avoid the need to parse the output.
Probably better to discuss via a Zoom call :)

**Comment by yevgenypats** [comment]
> > I specifically built the cobra in such way that it should be possible to do so via `NewRootCmd` which is the main entry point so there shouldn't be any difference between this and an external process.
> 
> Is there a way to run isolated tests in a temp directory and not in the CWD? That should allow us to reduce dependencies between tests and run tests in parallel.
> 
> Also I think testing stdout is not necessarily bad in the case of CLIs as that's what the user experiences. We can leverage snapshot testing on the CLI output to avoid the need to parse the it. Probably better to discuss via a Zoom call :)

Yeah I think it should be possible to do in separate directories. Also, Im not a big fan of snapshot testing :) https://medium.com/@sapegin/whats-wrong-with-snapshot-tests-37fbe20dfe8e
but sure let's do zoom 📞

File: cli/integration_tests/gen_test.go [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
```
Comment (hermanschaaf) [inline cli/integration_tests/gen_test.go:51]:
```suggestion
			// sanity check only; we shouldn't be testing the full config here
			require.Equal(t, plugin.kind, config.Kind)
```

File: .github/workflows/test_unit.yml [hunk 7]
```diff
@@ -77,5 +76,5 @@ jobs:
 
       - name: Test
         if: steps.changed-files.outputs.any_changed == 'true'
-        run: go test -v ./...
+        run: go test -v ./... --tags unit
```
[the remark this answers is not in the document]
Comment (erezrokah) [reply .github/workflows/test_unit.yml:79]: Good point. I think I'll take the suggestion from https://github.com/cloudquery/cloudquery/issues/1803 to run all tests on all OSs

**Comment by erezrokah** [comment]
> Yeah I think it should be possible to do in separate directories. Also, Im not a big fan of snapshot testing :) https://medium.com/@sapegin/whats-wrong-with-snapshot-tests-37fbe20dfe8e

I'm familiar with the article and agree that in most cases (like testing Web apps with Jest) they don't provide much value.
I do think for CLIs they can provide benefits but lets discuss separately as it's out of scope for this PR and for v2 initial iteration. Opened a separate issue to discuss https://github.com/cloudquery/cloudquery/issues/1860

File: cli/cmd/generate_test.go [hunk 8]
```diff
@@ -68,11 +68,7 @@ spec:
 `
 
 func TestGenerate(t *testing.T) {
-	tmpdir, tmpErr := os.MkdirTemp("", "generate_test_*")
-	if tmpErr != nil {
-		t.Fatalf("failed to create temporary directory: %v", tmpErr)
-	}
-	defer os.RemoveAll(tmpdir)
+	tmpdir := t.TempDir()
```
Comment (hermanschaaf) [inline cli/cmd/generate_test.go:71]: Oh, nice, TIL that `testing.T` provides a `TempDir` function 🚀

**Review by hermanschaaf — APPROVED** [review]

**Review by yevgenypats — CHANGES_REQUESTED** [review]
Let's have a live discussion on the integration testing strategy?

File: cli/cmd/doc_test.go [hunk 9]
```diff
@@ -1,3 +1,5 @@
+//go:build unit
```
[the remark this answers is not in the document]
Comment (erezrokah) [reply cli/cmd/doc_test.go]: This code it outdated per @hermanschaaf's comment. We don't have a separate between unit tests and integration tests.
This PR introduced a separate workflow that run only the CLI unit tests on Windows and MacOS (yet to be completed):
https://github.com/cloudquery/cloudquery/pull/1848/files#diff-0dbaaba16effff4896e9054770ab112bf49ae5b1e41a70915abafeef253b9ca7R31

**Review by yevgenypats — CHANGES_REQUESTED** [review]
Let's discuss live but see comment here - https://github.com/cloudquery/cloudquery/issues/1860#issuecomment-1259419465 

I don't think that we need to test/maintain output snapshot/predictability of a CLI

**Comment by erezrokah** [comment]
> I don't think that we need to test/maintain output snapshot/predictability of a CLI

To clarify, this PR does not introduce snapshot testing. I adds a workflow to run exiting CLI tests on Windows and Mac and changes the CLI tests so they can be executed in a temporary directory and not the CWD of the tests

**Comment by erezrokah** [comment]
Replaced by https://github.com/cloudquery/cloudquery/pull/2204

**Pull request closed without merging**